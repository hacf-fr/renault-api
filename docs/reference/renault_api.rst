Renault high-level API Reference
================================

.. contents::
    :local:
    :backlinks: none


Session
---------------

.. automodule:: renault_api.renault_session
   :members:

Client
--------------

.. automodule:: renault_api.renault_client
   :members:

Account
---------------

.. automodule:: renault_api.renault_account
   :members:

Vehicle
---------------

.. automodule:: renault_api.renault_vehicle
   :members:
