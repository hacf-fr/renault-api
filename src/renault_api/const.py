"""Constants for Renault API."""
CONF_COUNTRY = "country"
CONF_LOCALE = "locale"
CONF_GIGYA_APIKEY = "gigya-api-key"
CONF_GIGYA_URL = "gigya-root-url"
CONF_KAMEREON_APIKEY = "kamereon-api-key"
CONF_KAMEREON_URL = "kamereon-root-url"

PERMANENT_KEYS = [
    CONF_COUNTRY,
    CONF_LOCALE,
    CONF_GIGYA_APIKEY,
    CONF_GIGYA_URL,
    CONF_KAMEREON_APIKEY,
    CONF_KAMEREON_URL,
]


GIGYA_URL_EU = "https://accounts.eu1.gigya.com"
GIGYA_URL_US = "https://accounts.us1.gigya.com"
KAMEREON_APIKEY = "Ae9FDWugRxZQAGm3Sxgk7uJn6Q4CGEA2"
KAMEREON_URL_EU = "https://api-wired-prod-1-euw1.wrd-aws.com"
KAMEREON_URL_US = "https://api-wired-prod-1-usw2.wrd-aws.com"

LOCALE_BASE_URL = (
    "https://renault-wrd-prod-1-euw1-myrapp-one.s3-eu-west-1.amazonaws.com"
)

AVAILABLE_LOCALES = {
    "bg_BG": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3__3ER_6lFvXEXHTP_faLtq6eEdbKDXd9F5GoKwzRyZq37ZQ-db7mXcLzR1Jtls5sn",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "cs_CZ": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_oRlKr5PCVL_sPWUZdJ8c5NOl5Ej8nIZw7VKG7S9Rg36UkDszFzfHfxCaUAUU5or2",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "da_DK": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_5x-2C8b1R4MJPQXkwTPdIqgBpcw653Dakw_ZaEneQRkTBdg9UW9Qg_5G-tMNrTMc",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "de_DE": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_7PLksOyBRkHv126x5WhHb-5pqC1qFR8pQjxSeLB6nhAnPERTUlwnYoznHSxwX668",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "de_AT": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3__B4KghyeUb0GlpU62ZXKrjSfb7CPzwBS368wioftJUL5qXE0Z_sSy0rX69klXuHy",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "de_CH": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_UyiWZs_1UXYCUqK_1n7l7l44UiI_9N9hqwtREV0-UYA_5X7tOV-VKvnGxPBww4q2",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "en_GB": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_e8d4g4SE_Fo8ahyHwwP7ohLGZ79HKNN2T8NjQqoNnk6Epj6ilyYwKdHUyCw3wuxz",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "en_IE": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_Xn7tuOnT9raLEXuwSI1_sFFZNEJhSD0lv3gxkwFtGI-RY4AgiePBiJ9EODh8d9yo",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "es_ES": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_DyMiOwEaxLcPdBTu63Gv3hlhvLaLbW3ufvjHLeuU8U5bx3zx19t5rEKq7KMwk9f1",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "es_MX": {
        CONF_GIGYA_URL: GIGYA_URL_US,
        CONF_GIGYA_APIKEY: "3_BFzR-2wfhMhUs5OCy3R8U8IiQcHS-81vF8bteSe8eFrboMTjEWzbf4pY1aHQ7cW0",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_US,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "fi_FI": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_xSRCLDYhk1SwSeYQLI3DmA8t-etfAfu5un51fws125ANOBZHgh8Lcc4ReWSwaqNY",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "fr_FR": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_4LKbCcMMcvjDm3X89LU4z4mNKYKdl_W0oD9w-Jvih21WqgJKtFZAnb9YdUgWT9_a",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "fr_BE": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_ZK9x38N8pzEvdiG7ojWHeOAAej43APkeJ5Av6VbTkeoOWR4sdkRc-wyF72HzUB8X",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "fr_CH": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_h3LOcrKZ9mTXxMI9clb2R1VGAWPke6jMNqMw4yYLz4N7PGjYyD0hqRgIFAIHusSn",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "fr_LU": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_zt44Wl_wT9mnqn-BHrR19PvXj3wYRPQKLcPbGWawlatFR837KdxSZZStbBTDaqnb",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "hr_HR": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_HcDC5GGZ89NMP1jORLhYNNCcXt7M3thhZ85eGrcQaM2pRwrgrzcIRWEYi_36cFj9",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "hu_HU": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_nGDWrkSGZovhnVFv5hdIxyuuCuJGZfNmlRGp7-5kEn9yb0bfIfJqoDa2opHOd3Mu",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "it_IT": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_js8th3jdmCWV86fKR3SXQWvXGKbHoWFv8NAgRbH7FnIBsi_XvCpN_rtLcI07uNuq",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "it_CH": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_gHkmHaGACxSLKXqD_uDDx415zdTw7w8HXAFyvh0qIP0WxnHPMF2B9K_nREJVSkGq",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "nl_NL": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_ZIOtjqmP0zaHdEnPK7h1xPuBYgtcOyUxbsTY8Gw31Fzy7i7Ltjfm-hhPh23fpHT5",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "nl_BE": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_yachztWczt6i1pIMhLIH9UA6DXK6vXXuCDmcsoA4PYR0g35RvLPDbp49YribFdpC",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "no_NO": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_QrPkEJr69l7rHkdCVls0owC80BB4CGz5xw_b0gBSNdn3pL04wzMBkcwtbeKdl1g9",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "pl_PL": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_2YBjydYRd1shr6bsZdrvA9z7owvSg3W5RHDYDp6AlatXw9hqx7nVoanRn8YGsBN8",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "pt_PT": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3__afxovspi2-Ip1E5kNsAgc4_35lpLAKCF6bq4_xXj2I2bFPjIWxAOAQJlIkreKTD",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "ro_RO": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_WlBp06vVHuHZhiDLIehF8gchqbfegDJADPQ2MtEsrc8dWVuESf2JCITRo5I2CIxs",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "ru_RU": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_N_ecy4iDyoRtX8v5xOxewwZLKXBjRgrEIv85XxI0KJk8AAdYhJIi17LWb086tGXR",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "sk_SK": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_e8d4g4SE_Fo8ahyHwwP7ohLGZ79HKNN2T8NjQqoNnk6Epj6ilyYwKdHUyCw3wuxz",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "sl_SI": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_QKt0ADYxIhgcje4F3fj9oVidHsx3JIIk-GThhdyMMQi8AJR0QoHdA62YArVjbZCt",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "sv_SE": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_EN5Hcnwanu9_Dqot1v1Aky1YelT5QqG4TxveO0EgKFWZYu03WkeB9FKuKKIWUXIS",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
}
