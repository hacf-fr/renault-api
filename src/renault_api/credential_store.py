"""Kamereon client for interaction with Renault servers."""
import json
import os
from typing import Dict
from typing import List
from typing import Optional

import jwt

from renault_api.const import PERMANENT_KEYS
from renault_api.credential import Credential
from renault_api.credential import JWTCredential


class CredentialStore:
    """Credential store."""

    def __init__(self) -> None:
        """Initialise the credential store."""
        self._store: Dict[str, Credential] = {}

    def __getitem__(self, name: str) -> Credential:
        """Get a credential the credential store."""
        if name in list(self._store.keys()):
            cred = self._store[name]
            if not cred.has_expired():
                return cred
        raise KeyError(name)

    def get(self, name: str) -> Optional[Credential]:
        """Get a credential the credential store."""
        if name in list(self._store.keys()):
            cred = self._store[name]
            if not cred.has_expired():
                return cred
        return None

    def get_value(self, name: str) -> Optional[str]:
        """Get a credential value from the credential store."""
        if name in list(self._store.keys()):
            cred = self._store[name]
            if not cred.has_expired():
                return cred.value
        return None

    def __delitem__(self, name: str) -> None:
        """Remove a credential from the credential store."""
        del self._store[name]
        self._write()

    def __setitem__(self, name: str, value: Credential) -> None:
        """Add a credential to the credential store."""
        if not isinstance(name, str):  # pragma: no cover
            raise TypeError("`name` must be a string")
        if not isinstance(value, Credential):  # pragma: no cover
            raise TypeError("`value` must be a Credential")

        self._store[name] = value
        self._write()

    def __contains__(self, name: str) -> bool:
        """Check if a credential is in the credential store."""
        if name in self._store:
            cred = self._store[name]
            if not cred.has_expired():
                return True
        return False

    def _write(self) -> None:
        """Writes the content to fixed storage."""
        pass

    def clear(self) -> None:
        """Remove all non-permanent keys from credential store."""
        for key in list(self._store.keys()):
            if key not in PERMANENT_KEYS:
                del self._store[key]
        self._write()

    def clear_keys(self, to_delete: List[str]) -> None:
        """Remove specified keys from credential store."""
        for key in list(self._store.keys()):
            if key in to_delete:
                del self._store[key]
        self._write()


class CredentialEncoder(json.JSONEncoder):
    """Custom encoder for Credential class."""

    def default(self, obj: Credential) -> str:
        """Store the value."""
        return obj.value


class FileCredentialStore(CredentialStore):
    """Credential store with items stored in a file."""

    def __init__(self, store_location: str) -> None:
        """Initialise the credential store."""
        super().__init__()
        self._store_location = store_location
        self._read()

    def _read(self) -> None:
        """Read data from store location."""
        if not os.path.exists(self._store_location):
            return
        with open(self._store_location) as json_file:
            data = json.load(json_file)
            for key, value in data.items():
                if key == "gigya_jwt":
                    try:
                        self[key] = JWTCredential(value)
                    except jwt.ExpiredSignatureError:  # pragma: no cover
                        pass
                else:
                    self[key] = Credential(value)

    def _write(self) -> None:
        """Write data to store location."""
        dirname = os.path.dirname(self._store_location)
        if not os.path.isdir(dirname):
            os.makedirs(dirname)
        with open(self._store_location, "w") as json_file:
            json.dump(self._store, json_file, cls=CredentialEncoder)
