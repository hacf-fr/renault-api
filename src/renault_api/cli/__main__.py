"""Command-line interface."""
import errno
import json
import logging
import os
from datetime import datetime
from io import TextIOWrapper
from typing import Any
from typing import Dict
from typing import Optional

import aiohttp
import click
from click.core import Context

from . import helpers
from . import renault_account
from . import renault_client
from . import renault_settings
from . import renault_vehicle
from .charge import commands as charge_commands
from .hvac import commands as hvac_commands
from renault_api.credential_store import FileCredentialStore


_WARNING_DEBUG_ENABLED = (
    "Debug output enabled. Logs may contain personally identifiable "
    "information and account credentials! Be sure to sanitise these logs "
    "before sending them to a third party or posting them online."
)


def _check_for_debug(debug: bool, log: bool) -> None:
    """Renault CLI."""
    if debug or log:
        renault_log = logging.getLogger("renault_api")
        renault_log.setLevel(logging.DEBUG)

        if log:
            # create directory
            try:
                os.makedirs("logs")
            except OSError as e:  # pragma: no cover
                if e.errno != errno.EEXIST:
                    raise

            # create formatter and add it to the handlers
            formatter = logging.Formatter(
                "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
            )

            # create file handler which logs even debug messages
            fh = logging.FileHandler(f"logs/{datetime.today():%Y-%m-%d}.log")
            fh.setLevel(logging.DEBUG)
            fh.setFormatter(formatter)

            # And enable our own debug logging
            renault_log.addHandler(fh)

        if debug:
            logging.basicConfig()

        renault_log.warning(_WARNING_DEBUG_ENABLED)


@click.group()
@click.version_option()
@click.option("--debug", is_flag=True, help="Display debug traces.")
@click.option("--log", is_flag=True, help="Log debug traces to file.")
@click.option("--json", is_flag=True, help="Return data as JSON")
@click.option("--locale", default=None, help="API locale (eg. fr_FR)")
@click.option(
    "--account",
    default=None,
    help="Kamereon account ID to use",
)
@click.option("--vin", default=None, help="Vehicle VIN to use")
@click.pass_context
def main(
    ctx: Context,
    *,
    debug: bool,
    log: bool,
    json: bool,
    locale: Optional[str] = None,
    account: Optional[str] = None,
    vin: Optional[str] = None,
) -> None:
    """Main entry point for the Renault CLI."""
    ctx.ensure_object(dict)
    ctx.obj["credential_store"] = FileCredentialStore(
        os.path.expanduser(renault_settings.CREDENTIAL_PATH)
    )
    _check_for_debug(debug, log)
    ctx.obj["json"] = json
    if locale:
        ctx.obj["locale"] = locale
    if account:
        ctx.obj["account"] = account
    if vin:  # pragma: no branch
        ctx.obj["vin"] = vin


main.add_command(charge_commands.charge)
main.add_command(hvac_commands.hvac)


@main.command()
@click.pass_obj
@helpers.coro_with_websession
async def accounts(
    ctx_data: Dict[str, Any],
    *,
    websession: aiohttp.ClientSession,
) -> None:
    """Display list of accounts."""
    await renault_client.display_accounts(websession, ctx_data)


@main.command()
@click.option("--user", prompt=True)
@click.option("--password", prompt=True, hide_input=True)
@click.pass_obj
@helpers.coro_with_websession
async def login(
    ctx_data: Dict[str, Any],
    *,
    user: str,
    password: str,
    websession: aiohttp.ClientSession,
) -> None:
    """Login to Renault."""
    await renault_client.login(websession, ctx_data, user, password)


@main.command()
def reset() -> None:
    """Clear all credentials/settings from the credential store."""
    renault_settings.reset()


@main.command()
@click.option("--locale", default=None, help="API locale (eg. fr_FR)")
@click.option(
    "--account", default=None, help="Kamereon account ID to use for future calls"
)
@click.option("--vin", default=None, help="Vehicle VIN to use for future calls")
@click.pass_obj
@helpers.coro_with_websession
async def set(
    ctx_data: Dict[str, Any],
    *,
    locale: Optional[str] = None,
    account: Optional[str] = None,
    vin: Optional[str] = None,
    websession: aiohttp.ClientSession,
) -> None:
    """Store specified settings into credential store."""
    await renault_settings.set_options(websession, ctx_data, locale, account, vin)


@main.command()
@click.pass_obj
def settings(ctx_data: Dict[str, Any]) -> None:
    """Display the current configuration keys."""
    renault_settings.display_settings(ctx_data)


@main.command()
@click.pass_obj
@helpers.coro_with_websession
async def status(
    ctx_data: Dict[str, Any],
    *,
    websession: aiohttp.ClientSession,
) -> None:
    """Display vehicle status."""
    await renault_vehicle.display_status(websession, ctx_data)


@main.command()
@click.pass_obj
@helpers.coro_with_websession
async def vehicles(
    ctx_data: Dict[str, Any],
    *,
    websession: aiohttp.ClientSession,
) -> None:
    """Display list of vehicles."""
    await renault_account.display_vehicles(websession, ctx_data)


@main.command()
@click.pass_obj
@helpers.coro_with_websession
async def vehicle(
    ctx_data: Dict[str, Any],
    *,
    websession: aiohttp.ClientSession,
) -> None:
    """Display vehicle details."""
    await renault_vehicle.display_vehicle(websession, ctx_data)


@main.command()
@click.pass_obj
@helpers.coro_with_websession
async def contracts(
    ctx_data: Dict[str, Any],
    *,
    websession: aiohttp.ClientSession,
) -> None:
    """Display vehicle contracts."""
    await renault_vehicle.display_contracts(websession, ctx_data)


@main.group()
def http() -> None:
    """Raw HTTP."""
    pass


@http.command(name="get")
@click.argument("endpoint")
@click.pass_obj
@helpers.coro_with_websession
async def http_get(
    ctx_data: Dict[str, Any],
    *,
    endpoint: str,
    websession: aiohttp.ClientSession,
) -> None:
    """Process HTTP GET request on endpoint."""
    await renault_client.http_request(websession, ctx_data, "GET", endpoint)


@http.command(name="post-file")
@click.argument("endpoint")
@click.argument("json-body", type=click.File("rb"))
@click.pass_obj
@helpers.coro_with_websession
async def http_post_file(
    ctx_data: Dict[str, Any],
    *,
    endpoint: str,
    json_body: TextIOWrapper,
    websession: aiohttp.ClientSession,
) -> None:
    """Process HTTP POST request on endpoint."""
    await renault_client.http_request(
        websession, ctx_data, "POST", endpoint, json.load(json_body)
    )


@http.command(name="post")
@click.argument("endpoint")
@click.argument("json-body")
@click.pass_obj
@helpers.coro_with_websession
async def http_post(
    ctx_data: Dict[str, Any],
    *,
    endpoint: str,
    json_body: str,
    websession: aiohttp.ClientSession,
) -> None:
    """Process HTTP POST request on endpoint."""
    await renault_client.http_request(
        websession, ctx_data, "POST", endpoint, json.loads(json_body)
    )


if __name__ == "__main__":  # pragma: no cover
    main(prog_name="renault-api")
