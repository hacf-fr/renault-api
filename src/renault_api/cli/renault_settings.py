"""Singletons for the CLI."""
import os
from textwrap import TextWrapper
from typing import Any
from typing import Dict
from typing import Optional

import aiohttp
import click
from tabulate import tabulate

from renault_api.const import CONF_LOCALE
from renault_api.credential import Credential
from renault_api.credential_store import CredentialStore
from renault_api.helpers import get_api_keys

CONF_ACCOUNT_ID = "accound-id"
CONF_VIN = "vin"

CREDENTIAL_PATH = "~/.credentials/renault-api.json"


async def set_options(
    websession: aiohttp.ClientSession,
    ctx_data: Dict[str, Any],
    locale: Optional[str],
    account: Optional[str],
    vin: Optional[str],
) -> None:
    """Set configuration keys."""
    credential_store: CredentialStore = ctx_data["credential_store"]
    if locale:
        # Ensure API keys are available
        api_keys = await get_api_keys(locale, websession=websession)

        credential_store[CONF_LOCALE] = Credential(locale)
        for k, v in api_keys.items():
            credential_store[k] = Credential(v)

    if account:
        credential_store[CONF_ACCOUNT_ID] = Credential(account)
    if vin:
        credential_store[CONF_VIN] = Credential(vin)


def display_settings(ctx_data: Dict[str, Any]) -> None:
    """Get the current configuration keys."""
    credential_store: CredentialStore = ctx_data["credential_store"]
    wrapper = TextWrapper(width=80)
    items = list(
        [key, "\n".join(wrapper.wrap(credential_store.get_value(key) or "-"))]
        for key in credential_store._store.keys()
    )
    click.echo(tabulate(items, headers=["Key", "Value"]))


def reset() -> None:
    """Clear all credentials/settings from the credential store."""
    try:
        os.remove(os.path.expanduser(CREDENTIAL_PATH))
    except FileNotFoundError:
        pass
