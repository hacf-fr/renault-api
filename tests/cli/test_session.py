"""Test cases for the __main__ module."""
import os

from aioresponses import aioresponses
from click.testing import CliRunner
from tests.const import TEST_ACCOUNT_ID
from tests.const import TEST_LOCALE
from tests.const import TEST_LOCALE_DETAILS
from tests.const import TEST_VIN

from renault_api.cli import __main__
from renault_api.cli.renault_settings import CONF_ACCOUNT_ID
from renault_api.cli.renault_settings import CONF_VIN
from renault_api.cli.renault_settings import CREDENTIAL_PATH
from renault_api.const import CONF_GIGYA_APIKEY
from renault_api.const import CONF_GIGYA_URL
from renault_api.const import CONF_KAMEREON_APIKEY
from renault_api.const import CONF_KAMEREON_URL
from renault_api.const import CONF_LOCALE
from renault_api.credential_store import FileCredentialStore


def test_set_locale(mocked_responses: aioresponses, cli_runner: CliRunner) -> None:
    """Test set locale."""
    credential_store = FileCredentialStore(os.path.expanduser(CREDENTIAL_PATH))
    for key in [
        CONF_LOCALE,
        CONF_GIGYA_APIKEY,
        CONF_GIGYA_URL,
        CONF_KAMEREON_APIKEY,
        CONF_KAMEREON_URL,
    ]:
        assert key not in credential_store

    result = cli_runner.invoke(__main__.main, f"set --locale {TEST_LOCALE}")
    assert result.exit_code == 0

    credential_store = FileCredentialStore(os.path.expanduser(CREDENTIAL_PATH))
    for key in [
        CONF_LOCALE,
        CONF_GIGYA_APIKEY,
        CONF_GIGYA_URL,
        CONF_KAMEREON_APIKEY,
        CONF_KAMEREON_URL,
    ]:
        assert key in credential_store

    assert "" == result.output


def test_set_account(mocked_responses: aioresponses, cli_runner: CliRunner) -> None:
    """Test set locale."""
    credential_store = FileCredentialStore(os.path.expanduser(CREDENTIAL_PATH))
    for key in [
        CONF_ACCOUNT_ID,
    ]:
        assert key not in credential_store

    result = cli_runner.invoke(__main__.main, f"set --account {TEST_ACCOUNT_ID}")
    assert result.exit_code == 0

    credential_store = FileCredentialStore(os.path.expanduser(CREDENTIAL_PATH))
    for key in [
        CONF_ACCOUNT_ID,
    ]:
        assert key in credential_store

    assert "" == result.output


def test_set_vin(mocked_responses: aioresponses, cli_runner: CliRunner) -> None:
    """Test set vin."""
    credential_store = FileCredentialStore(os.path.expanduser(CREDENTIAL_PATH))
    for key in [
        CONF_VIN,
    ]:
        assert key not in credential_store

    result = cli_runner.invoke(__main__.main, f"set --vin {TEST_VIN}")
    assert result.exit_code == 0

    credential_store = FileCredentialStore(os.path.expanduser(CREDENTIAL_PATH))
    for key in [
        CONF_VIN,
    ]:
        assert key in credential_store

    assert "" == result.output


def test_get_keys_succeeds(
    mocked_responses: aioresponses, cli_runner: CliRunner
) -> None:
    """It exits with a status code of zero."""
    result = cli_runner.invoke(__main__.main, f"set --locale {TEST_LOCALE}")
    assert result.exit_code == 0

    result = cli_runner.invoke(__main__.main, "settings")
    assert result.exit_code == 0

    expected_output = (
        "Key                Value\n"
        f"{'-'*17}  {'-'*66}\n"
        f"locale             {TEST_LOCALE}\n"
        f"{CONF_GIGYA_URL}     {TEST_LOCALE_DETAILS[CONF_GIGYA_URL]}\n"
        f"{CONF_GIGYA_APIKEY}      {TEST_LOCALE_DETAILS[CONF_GIGYA_APIKEY]}\n"
        f"{CONF_KAMEREON_URL}  {TEST_LOCALE_DETAILS[CONF_KAMEREON_URL]}\n"
        f"{CONF_KAMEREON_APIKEY}   {TEST_LOCALE_DETAILS[CONF_KAMEREON_APIKEY]}\n"
    )
    assert expected_output == result.output


def test_reset(mocked_responses: aioresponses, cli_runner: CliRunner) -> None:
    """Test set vin."""
    assert not os.path.exists(os.path.expanduser(CREDENTIAL_PATH))

    result = cli_runner.invoke(__main__.main, f"set --locale {TEST_LOCALE}")
    assert result.exit_code == 0
    assert os.path.exists(os.path.expanduser(CREDENTIAL_PATH))

    # Reset a first time - file should get deleted
    result = cli_runner.invoke(__main__.main, "reset")
    assert result.exit_code == 0
    assert not os.path.exists(os.path.expanduser(CREDENTIAL_PATH))

    # Reset a second time - make sure it doesn't error
    result = cli_runner.invoke(__main__.main, "reset")
    assert result.exit_code == 0
    assert not os.path.exists(os.path.expanduser(CREDENTIAL_PATH))
