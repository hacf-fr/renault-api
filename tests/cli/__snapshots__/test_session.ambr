# serializer version: 1
# name: test_get_keys_succeeds
  '''
  Key                Value
  -----------------  ------------------------------------------------------------------
  locale             fr_FR
  gigya-root-url     https://accounts.eu1.gigya.com
  gigya-api-key      3_4LKbCcMMcvjDm3X89LU4z4mNKYKdl_W0oD9w-Jvih21WqgJKtFZAnb9YdUgWT9_a
  kamereon-root-url  https://api-wired-prod-1-euw1.wrd-aws.com
  kamereon-api-key   YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J
  
  '''
# ---
